File: ProblemSolving/450DSA/JavaScript/src/binarytree/MirrorOfATree.js
/**
 * @author Anirudh Sharma
 * 
 * Given a binary tree, the task is to create a new binary tree which is a mirror image of
 * the given binary tree.
 */
const mirror = (root) => {
    // Base case
    if (root === null) {
        return null;
    }
    // Get left and right children of the node
    let left = mirror(root.left);
    let right = mirror(root.right);
    // Swap the nodes
    root.right = left;
    root.left = right;
    return root;
};

const inorderTraversal = (node) => {
    if (node === null) {
        return;
    }
    inorderTraversal(node.left);
    console.log(node.data + " ");
    inorderTraversal(node.right);
};

function Node(data, left, right) {
    this.data = (data === undefined ? 0 : data);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

const main = () => {
    let root = new Node(1);
    root.left = new Node(3);
    root.right = new Node(2);
    inorderTraversal(mirror(root));
    console.log();

    root = new Node(10);
    root.left = new Node(20);
    root.right = new Node(30);
    root.left.left = new Node(40);
    root.left.right = new Node(60);
    inorderTraversal(mirror(root));
    console.log();

    root = new Node(1);
    root.left = new Node(2);
    root.right = new Node(3);
    root.left.left = new Node(4);
    root.left.right = new Node(5);
    inorderTraversal(mirror(root));
    console.log();
};

main();

File: ProblemSolving/450DSA/Python/src/binarytree/MirrorOfATree.py
"""
@author Anirudh Sharma

Given a Binary Tree, find diameter of it.
The diameter of a tree is the number of nodes on the longest path between two end nodes in the tree.

Constraints:
1 <= Number of nodes <= 10000
1 <= Data of a node <= 1000
"""


def mirror(root):
    # Special case
    if root is None:
        return root
    # Get left and right children of the node
    left = mirror(root.left)
    right = mirror(root.right)
    # Swap the nodes
    root.right = left
    root.left = right
    return root


def inorderTraversal(node):
    if node is None:
        return node
    inorderTraversal(node.left)
    print(node.data, end=" ")
    inorderTraversal(node.right)


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


if __name__ == "__main__":
    root = Node(1)
    root.left = Node(3)
    root.right = Node(2)
    inorderTraversal(mirror(root))
    print("\n")

    root = Node(10)
    root.left = Node(20)
    root.right = Node(30)
    root.left.left = Node(40)
    root.left.right = Node(60)
    inorderTraversal(mirror(root))
    print("\n")

    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    inorderTraversal(mirror(root))
    print("\n")
